controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+
+class JupyterLab(Tool):
+    name = "jupyter-lab"
+
+    def release_name(self, username):
+        return f"{self.name}-{username}"
+
+    def deploy_params(self, user):
+        conf = settings.TOOLS[self.name]
+        return {
+            "Username": user.username.lower(),
+            "aws.iamRole": user.iam_role_name,
+            "toolsDomain": settings.TOOLS_DOMAIN,
+            "cookie_secret": secrets.token_hex(32),
+            "authProxy.auth0_domain": conf["domain"],
+            "authProxy.auth0_client_id": conf["client_id"],
+            "authProxy.auth0_client_secret": conf["client_secret"],
+        }
+
+
+class ToolDeployment(object):
+
+    def __init__(self, deployment, user):
+        self.deployment = deployment
+        self.user = user
+
+    @classmethod
+    def list(cls, user):
+        deployments = client.AppsV1Api().list_namespaced_deployment(
+            user.k8s_namespace,
+        )
+        return [
+            ToolDeployment(deployment, user) for deployment in deployments.items
+        ]
+
+    @property
+    def name(self):
+        return self.deployment.metadata.labels["app"]
+
+    @property
+    def username(self):
+        return self.user.username.lower()
+
+    @property
+    def url(self):
+        return f"https://{self.username}-{self.name}.{settings.TOOLS_DOMAIN}"
+
+    @property
+    def pods(self):
+        pods = client.CoreV1Api().list_namespaced_pod(
+            self.user.k8s_namespace,
+            label_selector=f"app={self.name}",
+        )
+        return [Pod(pod) for pod in pods.items]
+
+    @property
+    def available(self):
+        return any(pod.status["phase"] == "Running" for pod in self.pods)
+
+    @property
+    def idled(self):
+        return self.deployment.metadata.labels.get(IDLED) == "true"
+
+    def restart(self):
+        return client.AppsV1Api().delete_collection_namespaced_replica_set(
+            namespace=self.user.k8s_namespaces,
+            label_selector=f"app={self.name}",
+            watch=True,
+        )
+
+
+class Pod(object):
+
+    def __init__(self, pod):
+        self.pod = pod
+
+    @property
+    def status(self):
+
+        if self.pod.status.container_status:
+            state = self.pod.status.container_status[0].state
+
+            if state.waiting:
+                return {"phase": "waiting", "reason": state.waiting.reason}
+
+            if state.terminated:
+                if not state.terminated.reason:
+                    if state.terminated.signal:
+                        reason = f"Signal:{state.terminated.signal}"
+                    else:
+                        reason = f"ExitCode:{state.terminated.exit_code}"
+                return {"phase": "terminated", "reason": reason}
+
+        return {"phase": self.pod.status.phase}
+
+    @property
+    def display_status(self):
+        status = self.status
+        reason = f":{status['reason']}" if status.get("reason") else ""
+        return f"{status['phase']}{reason}"
```

Comment:
#### 1e. Refactor Tools code
>Lines 209 - 237

The `Pod` class has been copied from the frontend Node webapp. It provides a property to calculate the status of a Kubernetes Pod, which is used to determine the status of a tool Deployment.

[Next →]()